File: question234/palindrome.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// had trouble with the odd cases
// 34m 
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if (!head->next) return true; 

        stack<int> data; 
        while (head) {
            if (!data.empty()) {
                if (data.top() != head->val) {
                    data.push(head->val); 
                } else {
                    data.pop(); 
                }
            }
            head = head->next; 
        }
        if ((data.empty()) || ((data.size() != 1) && (data.size() % 2 != 0))) return true; 
        return false; 
    }
};

File: question125/palindrome.py
# done in 15m
class Solution(object):
    def isPalindrome(self, s):
        i, j = 0, len(s) - 1
        while i <= j:
            if s[i].isalnum() and s[j].isalnum():
                if s[i].lower() != s[j].lower(): 
                    return False
                i += 1; j -= 1 
            elif not(s[i].isalnum()) and s[j].isalnum(): 
                i += 1 
            elif s[i].isalnum() and not(s[j].isalnum()):
                j -= 1 
            else: 
                i += 1; j -= 1 

        return True 